### DISCUSSION TITLE
How to fetch the last saved value?

### DISCUSSION THREAD

**Discussion by AntyaDev** [microsoft/FASTER#610] [Q&A] [answered]
- How to fetch the last saved value?
- Does FASTER support some offset value as Kafka does? If it does, how can I get it?

**Comment by badrishc** [comment, marked as answer]
> How to fetch the last saved value?

I assume you are talking about `FasterLog` and want to get the last committed value. `FasterLog` is committed until `log.CommitedUntilAddress`. One option is to scan that last committed page and find the last committed record.

```cs
var PageSize = 1L << logSettings.PageSizeBits;
var committedUntilAddress = log.CommittedUntilAddress;
var committedPageStart = committedUntilAddress & ~(PageSize - 1);
if (committedPageStart == committedUntilAddress) // corner case: committed until page boundary
   committedPageStart -= PageSize;

using (var iter = log.Scan(committedPageStart, committedUntilAddress))
{
   // skip to the last record of this scan
}
```

> Does FASTER support some offset value as Kafka does? If it does, how can I get it?

Every enqueue is associated with an offset. `TryEnqueue` returns this logical address as an `out` param.

**Reply by AntyaDev** [reply]
Thanks @badrishc

**Comment by fdcastel** [comment]
@badrishc -- Does https://github.com/microsoft/FASTER/pull/861 improves this answer in any way?
